### PR TITLE
raftstore: adjust default log gc limit automatically

### LINKED ISSUE #12418: raft_log_gc_count_limit should be adjusted according to the actual region size
## Development Task

Now we are suppose to support using dynamic region size #11515, we also need to change the gc count limit and size limit dynamically.

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <hustmh@gmail.com>

### What is changed and how it works?

Issue Number: Close #12418

What's Changed:

```commit-message
Adjust default raft log gc limit according to region split size.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
Adjust default config of log gc limit according to config of region size automatically
```

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -2966,6 +2965,22 @@ impl TiKvConfig {
             );
         }
 
+        // Raft log gc limit must be 75% region size at least;
+        if self.raft_store.raft_log_gc_size_limit < self.coprocessor.region_split_size * 3 / 4 {
+            warn!(
```
Comment (BusyJay) [inline src/config.rs:2970]: How about making it `info`?

**Review by BusyJay — APPROVED** [review]

**Review by gengliqi — APPROVED** [review]

**Comment by BusyJay** [comment]
Will changing `region-split-size` also adjust `region-max-size`?

**Review by tonyxuqqi — APPROVED** [review]

**Comment by tonyxuqqi** [comment]
LGTM

**Comment by 5kbpers** [comment]
> Will changing `region-split-size` also adjust `region-max-size`?

No, it will throw an error if `region-max-size` < `region-split-size`. Need to adjust them automatically as well?

**Comment by BusyJay** [comment]
> No, it will throw an error if `region-max-size` < `region-split-size`. Need to adjust them automatically as well?

Yes, otherwise it's difficult to adjust configurations.

**Comment by 5kbpers** [comment]
> > No, it will throw an error if `region-max-size` < `region-split-size`. Need to adjust them automatically as well?
> 
> Yes, otherwise it's difficult to adjust configurations.

Done, PTAL again.

File: components/raftstore/src/coprocessor/config.rs [hunk 2]
```diff
@@ -98,16 +98,16 @@ impl Default for Config {
 impl Config {
     pub fn validate(&mut self) -> Result<()> {
         if self.region_max_size.0 < self.region_split_size.0 {
```
Comment (BusyJay) [inline components/raftstore/src/coprocessor/config.rs:100]: How about making `region_max_size` as `Option<ReadableSize>`? So if it's not set, we can give it a correct number.

File: components/raftstore/src/coprocessor/config.rs [hunk 3]
```diff
@@ -98,16 +98,16 @@ impl Default for Config {
...
-            info!("region_max_size {:?} < region_split_size {:?}, adjust region_split_size to 125% region_max_size automatically",
-                self.region_max_size,
-                self.region_split_size,
+            info!(
+                "region_max_size {:?} < region_split_size {:?}, adjust region_split_size to 125% region_max_size automatically",
+                self.region_max_size, self.region_split_size,
             );
             self.region_max_size = self.region_split_size / 2 * 3;
```
Comment (BusyJay) [inline components/raftstore/src/coprocessor/config.rs:105]: It doesn't match the log.

Comment (5kbpers) [reply components/raftstore/src/coprocessor/config.rs:105]: Fixed

File: components/raftstore/src/coprocessor/config.rs (hunk 2 above)
Comment (5kbpers) [reply components/raftstore/src/coprocessor/config.rs]: Changed, PTAL again.

File: src/config.rs [hunk 4]
```diff
@@ -2980,9 +2979,30 @@ impl TiKvConfig {
...
 
+        #[cfg(not(any(test, feature = "textexport")))]
+        self.raft_log_gc_limit_adjust();
+
         Ok(())
     }
 
+    pub fn raft_log_gc_limit_adjust(&mut self) {
+        // Raft log gc limit must be 75% region size at least;
+        if self.raft_store.raft_log_gc_size_limit < self.coprocessor.region_split_size * 3 / 4 {
```
Comment (BusyJay) [inline src/config.rs:2990]: How about also using `Option`? This should not be a hard limit.

Comment (5kbpers) [reply src/config.rs:2990]: Too many changes were introduced previously, will try again if it was needed.

Comment (BusyJay) [reply src/config.rs:2990]: User should be allowed to use a smaller value for aggressive disk space reclaim.

**Review by BusyJay — APPROVED** [review]

**Comment by BusyJay** [comment]
You may also need to calculate default region-split-key based on region-split-size

**Comment by 5kbpers** [comment]
> You may also need to calculate default region-split-key based on region-split-size

Done, PTAL.

**Review by BusyJay — APPROVED** [review]
So many cases failed.

**Pull request merged**